### PR TITLE
*: fix vulkan packages deps

### PR DESCRIPTION
**Pull request by truboxl**

### CODE REVIEW

File: packages/libncnn/build.sh
Comment (xtkoba) [inline packages/libncnn/build.sh:37]: I doubt if this works as intended for submodules. Please use the method taken in `c-toxcore` or `telegram-bot-api` package. Also I recommend checking the SHA256 sum of the source tree.

File: packages/vulkan-loader-android/build.sh
Comment (licy183) [inline packages/vulkan-loader-android/build.sh:9]: I don't think this is a proper solution. Android's vulkan loader only supports Android Native Window, and doesn't support x11 or wayland. Can you explain why you want to let it provide `vulkan-loader-generic`?

**Review by licy183 — CHANGES_REQUESTED** [review]

File: packages/vulkan-loader-android/build.sh
Comment (truboxl) [reply packages/vulkan-loader-android/build.sh:9]: This is to match vulkan-loader-generic. This is the same for nodejs and nodejs-lts which will conflict each other file wise. The same can be said why vulkan-loader-generic should provide vulkan-loader-android when there is no Android Native Window? See termux-pacman issue #15531 not for the reason you think. This should have been done at each other a long time ago anyway.

If a package must use vulkan-loader-generic (strictly for X11 or Wayland features) and DT_NEEDED has libvukan.so / libvulkan.so.1, then the vulkan-loader-generic should already be listed in TERMUX_PKG_DEPENDS during package creation.

**Otherwise no package should ever runtime depends on vulkan-loader-generic and vulkan-loader-android** as libvulkan.so comes by default, see libandroid issue #16902 and vulkan-loader-android issue https://github.com/termux/termux-packages/pull/11422#issuecomment-1211494116

This also means all your mesa-vulkan-icd-* has wrong TERMUX_SUBPKG_DEPENDS and I will review accordingly #17112 

By the way, vulkan-loader virtual package is a mistake. See #17005 for an example of build system overwrite each other by pulling both vulkan-loader-generic and vulkan-loader-android.

Comment (licy183) [reply packages/vulkan-loader-android/build.sh:9]: > This is to match vulkan-loader-generic.

> This also means all your mesa-vulkan-icd-* has wrong TERMUX_SUBPKG_DEPENDS and I will review accordingly #17112

I haven't noticed it yet. I still don't think this is a proper way. `vulkan-loader-android` and `vulkan-loader-generic` have different functions and I don't think they can `PROVIDE` each other because they have considerable functional differences. The functions between them overlap in a basic vulkan loader, but the other functions are completely different.

> If a package must use vulkan-loader-generic (strictly for X11 or Wayland features) and DT_NEEDED has libvukan.so / libvulkan.so.1, then the vulkan-loader-generic should already be listed in TERMUX_PKG_DEPENDS during package creation.

> Otherwise no package should ever runtime depends on vulkan-loader-generic and vulkan-loader-android as libvulkan.so comes by default.

> By the way, vulkan-loader virtual package is a mistake.

If `vulkan-loader-android` comes by default because it exists in `/system/lib64`, is it necessary to have it in our repo?
And if we both have `vulkan-loader-android` and `vulkan-loader-generic` in our repo, I don't think `vulkan-loader` package is a mistake. It at least provides an alternative between `vulkan-loader-android` and `vulkan-loader-generic`.

> See #17005 for an example of build system overwrite each other by pulling both vulkan-loader-generic and vulkan-loader-android.

This is a bug of our building system. Our building system doesn't support `or` logic when parsing `TERMUX_(SUB)PKG_(BUILD_)DEPENDS`. Actually only one package should be pulled and extracted.

Currently we have three type of packages.

(1) Packages that should only uses `vulkan-loader-generic`: all of vulkan ICDs (`mesa-vulkan-icd-*`, `swiftshader`)

(2) Packages that should only uses `vulkan-loader-android`: (I don't know what exactly they are, but at least) `libncnn`, `fastfetch`

(3) Packages that uses either of `libvulkan.so`: `vulkan-tools`

IMO, packages in (1) should have `vulkan-loader-generic` in its `DEPENDS`, packages in (2) should have `vulkan-loader-android` in its `DEPENDS` and packages in (3) should have `vulkan-loader` in its `DEPENDS`. And this is what is being done now.

Ping: @Maxython , as the maintainer of termux-pacman

Comment (truboxl) [reply packages/vulkan-loader-android/build.sh:9]: vulkan-loader-android is strictly a build dependency package and as such should be used only in TERMUX_PKG_BUILD_DEPENDS. (2) and (3) will follow this.

TERMUX_PKG_DEPENDS should be reserved only for libraries that actually linked to libvulkan.so and none of the (1) fulfil this requirement. They should use TERMUX_PKG_RECOMMENDS vulkan-loader-generic if possible.

Eg: clinfo and clpeak links to libOpenCL.so and therefore shall TERMUX_PKG_DEPENDS on ocl-icd.

Comment (licy183) [reply packages/vulkan-loader-android/build.sh:9]: > TERMUX_PKG_DEPENDS should be reserved only for libraries that actually linked to libvulkan.so and none of the (1) fulfil this requirement. They should use TERMUX_PKG_RECOMMENDS vulkan-loader-generic if possible.

ICDs without `vulkan-loader-generic` is totally useless, so I still don't think it is a proper way to change `vulkan-loader-generic` from `TERMUX_PKG_DEPENDS` to `TERMUX_PKG_RECOMMENDS`. Some packages like `clang` still depends on `lld`, and it doesn't link against it either. IMO, `TERMUX_PKG_DEPENDS` should not only contain the packages that it will link against with, but also contain some packages that make this package work properly.

Following is how other distributions solve this problem:
In Ubuntu/Debian, `mesa-vulkan-drivers` depends on `libvulkan1`, and `libvulkan1` recommends `mesa-vulkan-drivers` .
In Fedora, `vulkan-drivers` requires (the same as depends) on `vulkan` (provided by `vulkan-loader`), and `vulkan-loader` recommends `vulkan-drivers`.
In Arch Linux, `vulkan-driver` doesn't depend on `vulkan-icd-loader` , and `vulkan-icd-loader` optionally depends (I think it is the same as recommend) on `vulkan-driver`.

Comment (truboxl) [reply packages/vulkan-loader-android/build.sh]: I have reassign TERMUX_PKG_DEPENDS to vulkan-loader. The other Khronos packages will be done separately.

As noted with latest CI https://github.com/termux/termux-packages/actions/runs/5460341950/jobs/9937222083?pr=17297#step:6:683,
TERMUX_PKG_DEPENDS on vulkan-loader will pull in both vulkan-loader-android and vulkan-loader-generic leading to indeterministic behaviour

**Comment by truboxl** [comment]
Will merge this if no more reviews by tomorrow.

Packages that uses vulkan will depend on vulkan-loader when necessary. Will follow up with another PR for packages that uses the other loader.
vulkan-loader will apply TERMUX_PKG_ANTI_BUILD_DEPENDS to prevent pulling both loaders.

**Review by licy183 — APPROVED** [review]

**Pull request merged**
